File: ride/src/main/scala/com/shevchyk/ride/domain/RideTemplate.scala
package com.shevchyk.ride.domain

import com.shevchyk.core.domain.*
import zio.json.*
import java.time.{Instant, LocalTime}
import java.util.UUID
import com.github.f4b6a3.uuid.UuidCreator

case class RideTemplateId(value: UUID) derives JsonCodec

object RideTemplateId:
  def generate(): RideTemplateId = RideTemplateId(UuidCreator.getTimeOrderedEpoch())

enum RecurrencePattern derives JsonCodec:
  case DAILY, WEEKLY_MON, WEEKLY_TUE, WEEKLY_WED, WEEKLY_THU, WEEKLY_FRI, WEEKLY_SAT, WEEKLY_SUN, WEEKDAYS, CUSTOM

final case class RideTemplate(
    id: RideTemplateId,
    companyId: CompanyId,
    clientId: PersonId,
    creatorId: PersonId,
    name: String,
    fromAddress: String,
    fromLat: Option[Double] = None,
    fromLng: Option[Double] = None,
    toAddress: String,
    toLat: Option[Double] = None,
    toLng: Option[Double] = None,
    preferredDriverId: Option[PersonId] = None,
    notes: Option[String] = None,
    recurrencePattern: RecurrencePattern,
    recurrenceDays: Option[String] = None,
    pickupTime: LocalTime,
    isActive: Boolean = true,
    flightNumber: Option[String] = None,
    isAirportTransfer: Boolean = false,
    price: Option[BigDecimal] = None,
    createdAt: Instant = Instant.now(),
    updatedAt: Instant = Instant.now()
) derives JsonCodec

final case class CreateRideTemplateRequest(
    clientId: String,
    name: String,
    fromAddress: String,
    fromLat: Option[Double] = None,
    fromLng: Option[Double] = None,
    toAddress: String,
    toLat: Option[Double] = None,
    toLng: Option[Double] = None,
    preferredDriverId: Option[String] = None,
    notes: Option[String] = None,
    recurrencePattern: String,
    recurrenceDays: Option[String] = None,
    pickupTime: String,
    flightNumber: Option[String] = None,
    isAirportTransfer: Boolean = false,
    price: Option[Double] = None
) derives JsonCodec

final case class GenerateRidesRequest(
    fromDate: String,
    toDate: String
) derives JsonCodec

File: web/lib/modules/core/models/ride_template.dart
class RideTemplate {
  final String id;
  final String companyId;
  final String clientId;
  final String creatorId;
  final String name;
  final String fromAddress;
  final double? fromLat;
  final double? fromLng;
  final String toAddress;
  final double? toLat;
  final double? toLng;
  final String? preferredDriverId;
  final String? notes;
  final String? flightNumber;
  final String recurrencePattern;
  final String? recurrenceDays;
  final String pickupTime;
  final bool isActive;
  final bool isAirportTransfer;
  final double? price;

  const RideTemplate({
    required this.id,
    required this.companyId,
    required this.clientId,
    required this.creatorId,
    required this.name,
    required this.fromAddress,
    this.fromLat,
    this.fromLng,
    required this.toAddress,
    this.toLat,
    this.toLng,
    this.preferredDriverId,
    this.notes,
    this.flightNumber,
    required this.recurrencePattern,
    this.recurrenceDays,
    required this.pickupTime,
    this.isActive = true,
    this.isAirportTransfer = false,
    this.price,
  });

  factory RideTemplate.fromJson(Map<String, dynamic> json) {
    return RideTemplate(
      id: json['id'] ?? '',
      companyId: json['companyId'] ?? '',
      clientId: json['clientId'] ?? '',
      creatorId: json['creatorId'] ?? '',
      name: json['name'] ?? '',
      fromAddress: json['fromAddress'] ?? '',
      fromLat: json['fromLat']?.toDouble(),
      fromLng: json['fromLng']?.toDouble(),
      toAddress: json['toAddress'] ?? '',
      toLat: json['toLat']?.toDouble(),
      toLng: json['toLng']?.toDouble(),
      preferredDriverId: json['preferredDriverId'],
      notes: json['notes'],
      flightNumber: json['flightNumber'],
      recurrencePattern: json['recurrencePattern'] ?? 'Daily',
      recurrenceDays: json['recurrenceDays'],
      pickupTime: json['pickupTime'] ?? '08:00',
      isActive: json['isActive'] ?? true,
      isAirportTransfer: json['isAirportTransfer'] ?? false,
      price: json['price']?.toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'companyId': companyId,
      'clientId': clientId,
      'creatorId': creatorId,
      'name': name,
      'fromAddress': fromAddress,
      'fromLat': fromLat,
      'fromLng': fromLng,
      'toAddress': toAddress,
      'toLat': toLat,
      'toLng': toLng,
      'preferredDriverId': preferredDriverId,
      'notes': notes,
      'flightNumber': flightNumber,
      'recurrencePattern': recurrencePattern,
      'recurrenceDays': recurrenceDays,
      'pickupTime': pickupTime,
      'isActive': isActive,
      'isAirportTransfer': isAirportTransfer,
      'price': price,
    };
  }
}

class CreateRideTemplateRequest {
  final String name;
  final String clientId;
  final String fromAddress;
  final String toAddress;
  final String pickupTime;
  final String recurrencePattern;
  final String? recurrenceDays;
  final String? preferredDriverId;
  final String? notes;
  final String? flightNumber;
  final bool isAirportTransfer;
  final double? price;

  const CreateRideTemplateRequest({
    required this.name,
    required this.clientId,
    required this.fromAddress,
    required this.toAddress,
    required this.pickupTime,
    required this.recurrencePattern,
    this.recurrenceDays,
    this.preferredDriverId,
    this.notes,
    this.flightNumber,
    this.isAirportTransfer = false,
    this.price,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'clientId': clientId,
      'fromAddress': fromAddress,
      'toAddress': toAddress,
      'pickupTime': pickupTime,
      'recurrencePattern': recurrencePattern,
      'recurrenceDays': recurrenceDays,
      'preferredDriverId': preferredDriverId,
      'notes': notes,
      'flightNumber': flightNumber,
      'isAirportTransfer': isAirportTransfer,
      'price': price,
    };
  }
}
